types/index.d.ts

```diff
@@ -0,0 +1,166 @@
+  }>
+
+  // A view builds a virtual DOM node representation of the application state.
+  type View<S> = (state: State<S>) => VDOM<S>
+
+  // The subscriptions function manages a set of subscriptions.
+  type Subscriptions<S> = (state: State<S>) => Subscription<S>[]
+
+  // A subscription represents subscriber activity.
+  type Subscription<S, D = any> = boolean | undefined | SubscriberDescriptor<S, D> | Unsubscribe
+
+  // A subscriber reacts to subscription updates.
+  type SubscriberDescriptor<S, D> = [Subscriber<S, D>, Payload<D>]
+  type Subscriber<S, D> = (dispatch: Dispatch<S>, props?: Payload<D>) => void | Unsubscribe
+
+  // An unsubscribe function cleans up a canceled subscription.
+  type Unsubscribe = () => void
+
+  // Middleware allows for custom processing during dispatching.
+  type Middleware<S> = (dispatch: Dispatch<S>) => Dispatch<S>
+
+  // ---------------------------------------------------------------------------
+
+  // A dispatched action handles an event in the context of the current state.
+  type Dispatch<S> = (action: Action<S>, props?: Payload<any>) => void
+
+  // An action transforms existing state and/or wraps another action.
+  type Action<S, P = any> = ActionTransform<S, P> | ActionDescriptor<S, P>
+  type ActionTransform<S, P = any> = (state: State<S>, props?: Payload<P>) => StateFormat<S> | Action<S>
+  type ActionDescriptor<S, P> = [ActionTransform<S, P>, Payload<P>]
+
+  // A transform carries out the transition from one state to another.
+  type Transform<S, P = any> = (state: StateFormat<S>, props?: Payload<P>) => StateFormat<S>
+
+  // State can either be on its own or associated with effects.
+  type StateFormat<S> = State<S> | StateWithEffects<S>
+
+  // Application state is accessible in every view, action, and subscription.
+  type State<S> = S
+
+  // State can be associated with a list of effects to run.
+  type StateWithEffects<S, D = any> = [State<S>, ...(Effect<S, D> | RunnerDescriptor<S, D>)[]]
+
+  // An effect is an abstraction over an impure process.
+  type Effect<S, D = any> = (..._: any[]) => RunnerDescriptor<S, D>
+
+  // A runner is where side effects and any additional dispatching may occur.
+  type RunnerDescriptor<S, D = any> = [Runner<S, D>, Payload<D>]
+  type Runner<S, D> = (dispatch: Dispatch<S>, props?: Payload<D>) => void | Promise<void>
+
+  // A payload is data given to an action, effect, or subscription.
+  type Payload<P> = P
+
+  // ---------------------------------------------------------------------------
+
+  // A virtual DOM node represents an actual DOM element.
+  type VDOM<S> = {
+    readonly type: VDOMNodeType
+    readonly props: PropList<S>
+    readonly children: VNode<S>[]
+    node: MaybeNode
+    readonly tag: Tag<S>
+    readonly key: Key
+    memo?: PropList<S>
+    events?: Record<string, Action<S>>
+
+    // `_VDOM` is a guard property which gives us a way to tell `VDOM` objects
+    // apart from `PropList` object.
+    _VDOM: true
```

Comment:
Could you explain how this works? As you know, vnodes don't actually have a `_VDOM` property, but somehow this helps the TypeScript compiler infer something?

Reply:
Yeah, so TypeScript does structural typing, which means the props an object has acts as a distinguishing factor rather than what the name of the object type is, as would be the case in nominal type systems, like Java, C++, etc.

Reply:
So, this is only a trick, I see, since vnodes don't have a `_VDOM` prop. Okay!

Reply:
Exactly!  `_VDOM` takes advantage of how TypeScript works without forcing the user to do anything special with their code.